nixos/modules/services/networking/gns3.nix

```diff
@@ -0,0 +1,261 @@
+            Path to the TLS certificate file. This certificate will
+            be offered to, and may be verified by, clients.
+          '';
+        };
+
+        keyFile = lib.mkOption {
+          type = lib.types.nullOr lib.types.path;
+          default = null;
+          example = "/secrets/gns3.key";
+          description = lib.mdDoc "Private key file for the certificate.";
+        };
+      };
+
+      dynamips = {
+        enable = lib.mkEnableOption (lib.mdDoc ''Whether to enable Dynamips support.'');
+        package = lib.mkPackageOptionMD pkgs "dynamips" {};
+      };
+
+      ubridge = {
+        enable = lib.mkEnableOption (lib.mdDoc ''Whether to enable uBridge support.'');
+        package = lib.mkPackageOptionMD pkgs "ubridge" {};
+      };
+
+      vpcs = {
+        enable = lib.mkEnableOption (lib.mdDoc ''Whether to enable VPCS support.'');
+        package = lib.mkPackageOptionMD pkgs "vpcs" {};
+      };
+
+      docker = {
+        enable = lib.mkEnableOption (lib.mdDoc ''Whether to enable Docker support.'');
+      };
+
+      libvirtd = {
+        enable = lib.mkEnableOption (lib.mdDoc ''Whether to enable libvirtd support.'');
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages =
+      [ cfg.package ]
+      ++ lib.optional cfg.dynamips.enable cfg.dynamips.package
+      ++ lib.optional cfg.ubridge.enable cfg.ubridge.package
+      ++ lib.optional cfg.vpcs.enable cfg.vpcs.package;
+
+    security.wrappers.ubridge = lib.mkIf cfg.ubridge.enable {
+      owner = "root";
+      group = "root";
+      capabilities = "cap_net_raw,cap_net_admin+ep";
+      source = "${cfg.ubridge.package}/bin/ubridge";
+    };
+
+    # Docker support requires the Docker daemon to be running.
+    virtualisation.docker.enable = lib.mkIf cfg.docker.enable true;
+    virtualisation.libvirtd.enable = lib.mkIf cfg.libvirtd.enable true;
+
+    users.users.gns3 = {
+      name = "gns3";
+      group = "gns3";
+      description = "GNS3 server user";
+      isSystemUser = true;
+      home = "/var/lib/gns3";
+      extraGroups = lib.optional cfg.docker.enable "docker";
+    };
+
+    users.groups.gns3 = {};
+
+    systemd.services.gns3-server = let
+      commandArgs = "--config /var/lib/gns3/gns3_server.conf"
+        + " --host ${cfg.host}"
+        + " --pid /run/gns3/server.pid"
+        + " --port ${toString cfg.port}"
+        + lib.optionalString (cfg.log.file != null) " --log ${cfg.log.file}"
+        + lib.optionalString (cfg.log.debug) " --debug"
+        + lib.optionalString (cfg.tls.enable) " --ssl"
+        + lib.optionalString (cfg.tls.certFile != null) " --certfile ${cfg.tls.certFile}"
+        + lib.optionalString (cfg.tls.keyFile != null) " --certkey ${cfg.tls.keyFile}"
+      ;
+    in
+    {
+      description = "GNS3 Server";
+
+      after = [ "network.target" "network-online.target" ];
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ];
+
```

Comment:
There should be a config reload trigger based on `configFile` changes here, right?

Reply:
Added GNS3 server configuration file to `X-Reload-Triggers` in SystemD unit.